docs/design_proposals/artifact-dependencies.md

```diff
@@ -0,0 +1,356 @@
+# Supporting dependencies between build artifacts
+
+## Background
+Refer [this](https://tinyurl.com/skaffold-modules) document which presents the design goals for introducing the concept of _modules_ in Skaffold. A prerequisite to being able to define modules and supporting cross module dependency is to first consider the latest version of the skaffold config (_currently `v2beta8`_) as an implicit module and allowing dependencies between artifacts defined in it. The current document aims to capture the major code changes necessary for achieving this.
+
+>*Note: Omitted some details for brevity, assuming reader's familiarity with the skaffold [codebase](https://github.com/GoogleContainerTools/skaffold).*
+
+## Config schema
+
+We introduce an `ArtifactDependency` slice within `Artifact` in `config.go`
+```go
+type ArtifactDependency struct {
+	ImageName string `yaml:"image" yamltags:"required"`
+	Alias string `yaml:"alias,omitempty"`
+}
+```
+This allows us to define a build stanza like below where image `leeroy-app` requires image `simple-go-app`:
+```yaml
+build:
+ artifacts:
+   - image: simple-go-app
+   - image: leeroy-app
+     requires:
+       - image: simple-go-app
+         alias: BASE
+```
+Alias is a token that will be replaced with the image reference in the builder definition files. If no value is provided for `alias` then it defaults to the value of `image`.
+
+## Config validation
+
+We add two new validations to the [validation](https://github.com/GoogleContainerTools/skaffold/blob/10275c66a142719897894308b9e566953712a0fe/pkg/skaffold/schema/validation/validation.go#L37) package after the introduction of artifact dependencies:
+- Cyclic references among artifacts.
+  - We cannot have image `A` depend on image `B` depend on image `C` depend on image `A`.
+  - We run a simple depth first search cycle detection algorithm treating our `Artifact` slice like a directed graph- image `A` depending on image `B` implies a directed edge from `A` to `B`.
+- Unique artifact aliases.
+  - We ensure that within *each* artifact dependency slice the aliases are unique. 
+
+## Referencing dependencies
+
+### Docker builder
+
+The `docker` builder will use the `alias` of an `ArtifactDependency` as a build argument key.
+ 
+ ```yaml
+ build:
+  artifacts:
+    - image: simple-go-app
+    - image: leeroy-app
+      requires:
+        - image: simple-go-app
+          alias: BASE
+ ```
+
+Here `BASE=gcr.io/X/simple-go-app:<tag>@sha:<sha>` is passed as a `--build-arg` (or a buildKit parameter) when building `leeroy-app`.
+
+### Custom builder
+
+The `custom` builder will be supplied each `ArtifactDependency`'s `alias` and image reference as environment variables keyed on `alias`. So they can be easily referenced in user-defined build definitions.
```

Comment:
Define what is a valid alias as per the Docker builder

Reply:
all aliases have the same restrictions since they are either used as environment variable keys or build arg keys or template variables. So I've added it in the validation section above.